src/data_source/storage/sql.rs

```diff
@@ -1170,6 +1192,401 @@ where
     }
 }
 
+#[async_trait]
+impl UpdateStateData for SqlStorage {
+    type Error = QueryError;
+    async fn insert_merkle_nodes<
+        Proof: MembershipProof<E, I, T> + Send + Sync + Debug + 'static,
+        E: Element + Send + Sync + Serialize,
+        I: Index + Send + Sync + Serialize,
+        T: NodeValue + Send + Sync,
+    >(
+        &mut self,
+        name: &'static str,
+        proof: Proof,
+        traversal_path: Vec<usize>,
+        block_number: u64,
+    ) -> QueryResult<()> {
+        let pos = proof.index().clone();
+        // First node in the merkle path is always the leaf node for which membership proof is provided
```

Comment:
I think this could be `MerkleNode::Empty` in the case that an element was deleted from the tree by a state update (e.g. for the fee state, we delete an account when its balance reaches 0)